### ISSUE TITLE
client-sever protocol minor points

### ISSUE THREAD

**Issue by solyaris** [bwilcox-1234/ChatScript#67]
Hi Bruce, 

with the aim to spread CS among Rubysts :), I'm starting to work on a simple Ruby language CS client:
https://github.com/solyaris/rChatScript/

I followed info on the [ClientServer-Manual](https://github.com/bwilcox-1234/ChatScript/blob/master/WIKI/CLIENTS-AND-SERVERS/ChatScript-ClientServer-Manual.md). 

I have two MINOR notes: 

## (1) Testing for server presence

[documentation](https://github.com/bwilcox-1234/ChatScript/blob/master/WIKI/CLIENTS-AND-SERVERS/ChatScript-ClientServer-Manual.md#testing-the-server) say:

> If you send the message: null 1 null (that’s the null string user id, the string of the character “1” as bot id, and the null string message, the server will send back the string of [you have no user id]with no logging done and minimal load on the server. This constitutes an echo-test to prove the server is running.

In facts that seems untrue. 

here below some test using netcat: 
If I send the sequence  `null`  `1`  `null`  as stated in doc, 
I do not get answer and client hangs:
```
$ nc 127.0.0.1 1024
^@1^@

^C
$
```

instead I have success with  `null`  `null`  `null` (three nulls); 
in this case I have back a void string and the connection is closed by server:
```
$ nc 127.0.0.1 1024
^@^@^@
$
```

## (2) New Users

In section [chatscript-protocol](https://github.com/bwilcox-1234/ChatScript/blob/master/WIKI/CLIENTS-AND-SERVERS/ChatScript-ClientServer-Manual.md#chatscript-protocol) of above doc: 

> The third string is the message. 

>If the message is null, this is a start of new conversation. This MUST be the first thing you do with a new user. Ideally you do it whenever a new conversation is starting with that user which is how the system knows the old conversation ended. 

> Usually script will detect that this is the start of a new conversation and say something like hello or welcome back to indicate the two parties are starting up a new conversation, though through the history file the system may have a lot of information about what has gone on in prior conversations. As long as the user is connected to the webpage, for example, you wouldn’t send a startup message again. 

That's not fully clear to me.
As far as I tested, putting a `null` message for an **existing** user, that will trigger a rule like 
```
    %input<%userfirstline
```

in a possible introductions topic like that: 

```
topic: ~INTRODUCTIONS (~emogoodbye ~emohello ~emohowzit name here )

#!x will match on every return start of a conversation
t: ( %input<%userfirstline %input>0 )
  ^keep()
  [Welcome back] [Hello, again.][Glad you came back.][Hi.][Hi, again.]

#!x matches every time on startup of a new conversation
t: ( %input<%userfirstline )
  ^keep()
  Welcome to ChatScript.

#!x issued only once, then erased
t: Have you been here before?
  #! yes
  a: (~yes) Welcome back!
  #! no
  a: (~no) Then welcome to your first time.
```

May you please specify/detail the use case/scenario where the  `null` message  is useful ? 

Very welcome an update on documentation section: [chatscript-protocol](https://github.com/bwilcox-1234/ChatScript/blob/master/WIKI/CLIENTS-AND-SERVERS/ChatScript-ClientServer-Manual.md#chatscript-protocol) .

Thanks!
giorgio

**Comment by bwilcox-1234** [comment]
The null message indicates the start of a new conversation, regardless of old or new user. If you and I chatted a week ago and then we get together again today, while I have the history in my head of my last chat, this is a new chat and I presumably will start again with hello, how are you and such.  IF you don't do null, then the system will merely presume this is that same conversation. Without special scripting to check time, it has no way of knowing that you fell asleep for a week.

**Comment by solyaris** [comment]
thanks, 

# (2)
>Without special scripting to check time, it has no way of knowing that you fell asleep for a week.

btw, what's the solution you suggest to measure this time (time(now) - time(last volley) ?
may you give a script example if already done? 

# (1) the initial "issue"
May you confirm that correct sequence to check if CS server is up and running is  `null` `null` `null` (three nulls) ?

Afterward, I'd update myself the wiki document

thanks!

**Comment by bwilcox-1234** [comment]
I have rewritten the server alive check documentation.  As for time of last volley, it is trivial to set a permanent user variable with time and diff from that. eg
$mytime = %fulltime

and time lapse in seconds  is just  %fulltime - $mytime
